cluster-autoscaler/cloudprovider/konvoy/konvoy_manager.go

```diff
@@ -161,102 +152,84 @@ func (k *KonvoyManager) addNodeToNodeGroup(nodeGroup string) error {
-			}
+	for _, node := range result.Items {
+		nodeName := kubernetesNodeName(&node, k.provisioner)
+		if nodeName == name {
+			klog.V(2).Infof("Get node by name: %v", name)
+			return &node, nil
 		}
 	}
 	return nil, nil
 }
 
-func (k *KonvoyManager) RemoveNodeFromNodeGroup(nodeGroup string, node string) error {
-	konvoyNode, err := k.getNodeByName(node)
-	if konvoyNode == nil || err != nil {
-		klog.Warningf("Can't delete node %s from nodegroup %s. Node does not exist.", node, nodeGroup)
-		return nil
+// RemoveNodeFromNodeGroup marks given node for deletion and decreases target node group size
+// by one.
+func (k *KonvoyManager) RemoveNodeFromNodeGroup(nodeGroupName string, nodeName string) error {
+	kubernetesNode, err := k.getNodeByName(nodeName)
+	if kubernetesNode == nil || err != nil {
+		klog.Warningf("Can't delete node %s from nodegroup %s. Node does not exist.", nodeName, nodeGroupName)
+		return err
 	}
-	if konvoyNode.ObjectMeta.Labels[nodeGroupLabel] != nodeGroup {
-		return fmt.Errorf("can't delete node %s from nodegroup %s. Node is not in nodegroup", node, nodeGroup)
+	if kubernetesNode.ObjectMeta.Labels[nodeGroupLabel] != nodeGroupName {
+		return fmt.Errorf("can't delete node %s from nodegroup %s. Node is not in nodegroup", nodeName, nodeGroupName)
 	}
 
-	/*
-	  TODO: we could mark a node for deletion or delete it using Konvoy library
-	  if err := k.kubeClient.CoreV1().Nodes().Delete(node.Name, &metav1.DeleteOptions{}); err != nil {
-	    klog.Errorf("failed to delete node %s from Konvoy cluster, err: %v", node.Name, err)
-	  }
-	*/
-	for i := 0; i < numRetries; i++ {
-		// Decrease the size of the pool in Konvoy
-		konvoyCluster := &konvoyclusterv1beta1.KonvoyCluster{}
-		konvoyCluster.Name = k.clusterName
-		clusterNamespacedName := types.NamespacedName{
-			Namespace: "kommander",
-			Name:      konvoyCluster.Name,
-		}
-		err = k.dynamicClient.Get(context.Background(), clusterNamespacedName, konvoyCluster)
-		if err != nil {
-			klog.Warningf("Error retrieving the konvoy cluster: %v -- %v", konvoyCluster.Name, err)
-		}
-		for i, pool := range konvoyCluster.Spec.ProvisionerConfiguration.NodePools {
-			if pool.Name == nodeGroup {
-				konvoyCluster.Spec.ProvisionerConfiguration.NodePools[i].Count--
-			}
-		}
+	currentTargetSize, err := k.GetNodeGroupTargetSize(nodeGroupName)
+	if err != nil {
+		return err
+	}
 
-		if err = k.dynamicClient.Update(context.Background(), konvoyCluster); err != nil {
-			klog.Warningf("Error updating the konvoy cluster: %v -- %v", konvoyCluster.Name, err)
-			err = fmt.Errorf("Failed to delete node %s: %v", node, err)
-		} else {
-			klog.Infof("Konvoy cluster updated successfully: %v", konvoyCluster.Name)
-			return nil
-		}
+	kubernetesNode.Annotations[KonvoyNodeAnnotationKey] = time.Now().String()
+	if err := k.dynamicClient.Update(context.Background(), kubernetesNode); err != nil {
+		return err
 	}
-	return err
+
+	return k.setNodeGroupTargetSize(nodeGroupName, currentTargetSize-1)
```

Comment:
yep, the deletion should be done -1 to avoid creating issues while de-provisioning.